### DISCUSSION TITLE
Can run in browser? (vs. in Node)

### DISCUSSION THREAD

**Discussion by trevyn** [windicss/windicss#10] [Q&A] [answered]
For use-cases like an interactive playground that gets dynamically recompiled, is windicss able to run completely in the browser? (vs. in Node)

Thinking of hacking it into a Svelte playground, which I believe uses an in-browser version of Rollup.

**Comment by voorjaar** [comment, marked as answer]
I believe Typescript natively supports the compilation browser running, but I haven't tested it yet. As you can see, this project is only dependent on arg, a command-line parameter parser. In addition to this dependency, it is a simple JavaScript code, compiled into es5, which should directly support running in the browser.

**Comment by voorjaar** [comment]
I just tested it in the browser and it works. Although there are two warnings, I will fix it soon.

https://github.com/voorjaar/windicss-online-demo

For svelte, you need to add this plugin, [rollup-plugin-node-polyfills](https://github.com/ionic-team/rollup-plugin-node-polyfills).

**Reply by voorjaar** [reply]
Just do it, and good luck.

**Comment by voorjaar** [comment]
I just released windicss [v2.0.2](https://github.com/voorjaar/windicss/commit/1615fd856ae79590153834437ce5b9cb899757a5). All warnings have been fixed, I believe it works very well in browser now. [The demo](https://github.com/voorjaar/windicss-online-demo) also updated.

**Comment by voorjaar** [comment]
> For use-cases like an interactive playground that gets dynamically recompiled, is windicss able to run completely in the browser? (vs. in Node)
> 
> Thinking of hacking it into a Svelte playground, which I believe uses an in-browser version of Rollup.

How is your project progressing? I will be more than happy to help you if you need it.

**Reply by trevyn** [reply]
>How is your project progressing? I will be more than happy to help you if you need it.

It's going great, thanks! There's lots of moving parts, and the Tailwind bit is a small part of it: https://github.com/trevyn/turbo

I checked out https://github.com/voorjaar/windicss-online-demo, and noticed that it breaks with windicss 2.1.2 (`'Processor' is not exported by node_modules/windicss/index.mjs`), but works when locking `"windicss": "=2.0.2"`.

That aside, I am very much liking this direction — I didn't expect this level of granularity offered by `Processor`! That's a great demo.

Have you considered a test runner that runs a full build of `tailwindcss` with a given `tailwind.config.js` file, then runs `windicss` with the same config file, and diffs the results in some way? Or is that level of identicalness not a goal?

**Reply by voorjaar** [reply]
Since windicssv2.1 has slightly changed the interface, I haven't updated the demo yet. Compiling all tailwindcss utilities is feasible, as long as you exhaust all possible classNames. Already done this when testing.

https://github.com/voorjaar/windicss/blob/main/test/assets/testClasses.ts

**Reply by voorjaar** [reply]
It seems you want to do a bigger project, I think what I need is an online playground that combines tailwind and svelte.

**Reply by trevyn** [reply]
Ok, this should work:

```
git clone --branch experiments https://github.com/trevyn/MDsveX.git
cd MDsveX/packages/site
npm install
npm run dev
```

(MDsveX is not my project, see:

- https://mdsvex.com/playground
- https://github.com/pngwn/MDsveX
- https://github.com/pngwn/REPLicant
- https://www.youtube.com/watch?v=S3j1fLzC8_E

)

So I hacked the MDsveX site's playground to include Tailwind; after launching the site above, you should be able to use Svelte + Tailwind classes in that repl, e.g.:

`<span class="bg-red-500">hi</span>`

And it should work instantly.

This was a few weeks ago, when windicss was less mature, so I ended up switching to using the full development build of straight `tailwindcss` (you can see in the commit history that I tried windicss!), but it's pretty clear that windicss is a better solution.

**Reply by voorjaar** [reply]
Ok thanks I understand.

**Reply by voorjaar** [reply]
> > How is your project progressing? I will be more than happy to help you if you need it.
> 
> It's going great, thanks! There's lots of moving parts, and the Tailwind bit is a small part of it: https://github.com/trevyn/turbo
> 
> I checked out https://github.com/voorjaar/windicss-online-demo, and noticed that it breaks with windicss 2.1.2 (`'Processor' is not exported by node_modules/windicss/index.mjs`), but works when locking `"windicss": "=2.0.2"`.
> 
> That aside, I am very much liking this direction — I didn't expect this level of granularity offered by `Processor`! That's a great demo.
> 
> Have you considered a test runner that runs a full build of `tailwindcss` with a given `tailwind.config.js` file, then runs `windicss` with the same config file, and diffs the results in some way? Or is that level of identicalness not a goal?

https://voorjaar.github.io/svelte-windicss-preprocess/ try this

**Reply by trevyn** [reply]
Ooooh, that’s very nice.
